### DISCUSSION TITLE
Is it possible to generate a rename column migration?

### DISCUSSION THREAD

**Discussion by Incanus3** [ebean-orm/ebean#3521] [General]
Hi guys,
we just started using ebean-migrations a week back and were mostly successful (we did encounter two minor issues which I reported [here](https://github.com/ebean-orm/ebean-migration/issues)). The only thing we weren't able to get working are rename migrations. I looked (briefly, since I don't understand all the details) over the code in `ebean-migration` and `ebean-ddl-generator` and found that that there is indeed a `RenameColumn` class which is supported by `ModelContainer.applyChangeSet()` and `MTable.apply()` and there's even a test migration `6.0__renameColumn.model.xml` applied by `ModelContainerTest.apply_renameColumn()`, but I can't find a way to generate this kind of migration. Could you please offer some hints as for how this can be done?

BTW, I also tried a semi-manual approach - make the change in model, run `DbMigration.generateMigration()` which generates drop+create, delete the generated sql files if any (in my case it didn't generate any because the column is not nullable and we don't specify `DbDefault`), update the XML as follows
```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<migration xmlns="http://ebean-orm.github.io/xml/ns/dbmigration">
    <changeSet type="apply">
<!--        <addColumn tableName="SMLOUVA_LEGITO_LIST">-->
<!--            <column name="allow_users" type="varchar(3000)" notnull="true"/>-->
<!--        </addColumn>-->

        <renameColumn tableName="SMLOUVA_LEGITO_LIST" oldName="allowed_users" newName="allow_users" />
    </changeSet>
<!--    <changeSet type="pendingDrops">-->
<!--        <dropColumn columnName="allowed_users" tableName="SMLOUVA_LEGITO_LIST"/>-->
<!--    </changeSet>-->
</migration>
```
and rerun `generateMigration()` hoping that it will see that the sql files for the version are missing and finish generating them, but it just reports `DbMigration> no changes detected - no migration written`.

Cheers

**Comment by rPraml** [comment]
Hello @Incanus3,

as far as I remember, renaming is a pain and at least not supprted by all platforms. (depending on platform, you have to remove indices/constraints, do rename and recreate indices afterwards) 

We avoid renames whenever it is possible, but there is a simple workaround:

Just rename the field in java. Ebean will create the new column and drop the old (in two migration steps)
You can add `@DbMigration(postAdd = "update ${table} set ${column} = old_column_name)`
Then the statement is added to migration, that copies the data, before dropping them.
Yes, this is no real rename and causes lots of IO. But that's what we have now and it will work relaiable on all platforms

**Comment by Incanus3** [comment]
Hi @rPraml and thanks for the reply.

Well, in our case, there are no indices, nor constraints on this column, so a simple `alter table ... rename column <old> to <new>` should be sufficient. So, if I understand correctly, there is a `renameColumn` xml element allowed by the schema, and it seems to be supported by the `applyChangeSet()` method, but there's really no way to generate it?

Allow me a different question then: if I were to create the xml manually, is there a way to generate the sql files for it? Because currently if the model.xml file is there (even if it was generated by `DbMigration.generateMigration()`) but the sql files are not, they will not be generated on next run. We noticed this when we added some not-null columns, but forgot to add `@DbDefault` to them - the `model.xml` file was generated, but then the script failed, reporting that newly added not-null columns must have defaults (which makes sense of course), so the .sql files were never generated and even if we added the `@DbDefault` and reran the generation script, it didn't generate them, so we needed to delete the `model.xml` file and then rerun again. So the question is basically, is there a way to force ebean to generate the missing .sql files for given `model.xml` version?

**Reply by rPraml** [reply]
>is there a way to force ebean to generate the missing .sql files for given model.xml version?

At first glance it looks like there is no implementation in BaseTableDdl. So, IF you manage to force .sql generation from .xml, I doubt, that it will generate `rename column` scripts. (I did a quick search and found no code, that generate a "rename column" statement)

As said, ebean will generate an "add" and a "drop" in the XML, when you rename a field in java 
You can discard the generated SQLs and do the rename by hand.
What you should *not* do, is to discard the XML (editing the XML and replacing it by a rename is OK)
If you discard the XML, ebean looses the information, about the new column, as ebean does a replay of all migrations.

> but there's really no way to generate it?

No, how would you detect a rename:
Example 1:
```java
class Foo {
    String name;
    String lastName;
    int age;
}
class Foo {
    String firstName;
    String lastName;
    int age;
}
```
Example 2:
```java
class Foo {
    String name;
    String lastName;
    int age;
}
class Foo {
    String name;
    String lastName;
    int weight;
}
```
So where did I do a rename, and where did I remove one field and replace by an other ;)

We would need some meta information. e.g `@DbMigration(renameFrom="name") String firstName`
But this has no one implemented, yet.

The easiest way (for now) is to copy the data - or - replace create/drop by your own rename logic.

**Reply by Incanus3** [reply]
> We would need some meta information. e.g @DbMigration(renameFrom="name") String firstName

yes, this is exactly what I expected

> What you should not do, is to discard the XML (editing the XML and replacing it by a rename is OK)

yeah, I wasn't going to discard it, what I wanted to do was

1. generate xml & sql (with add+drop)
2. delete the sql files
3. edit the xml - change add+drop to `renameColumn`
4. regenerate sql files

that's why I was asking if there's a way to force ebean to regenerate the sql after changing the xml

> You can discard the generated SQLs and do the rename by hand.

ok, I'll do this then

Thanks again, your help is very much appreciated

**Reply by Incanus3** [reply]
oh actually, now I've read this once more and noticed the "by hand" I missed before. I don't really wanna do it "by hand" on every target db, but I should be able to edit the generated sql files, no?

**Reply by rPraml** [reply]
Sorry, yes I mean, edit the generated file "by hand". So remove the add colum/drop colum statement and replace it with a rename statement

**Comment by rbygrave** [comment]
> but I can't find a way to generate this kind of migration. Could you please offer some hints as for how this can be done?

Migrations that would Rename a table or column are not automatically generated. For 2 reasons:
1. We'd need some sort of `@DbMigration(renameFrom="name")` and that doesn't exist
2. Rename migrations are NOT SAFE to run against a running application / they are BREAKING CHANGE migrations and this means that the only way to run them is to ensure the application is stopped or not running.

So if you want to do RENAME migrations we want people to understand that they are "Breaking migrations" and that they need to do that via scheduling downtime / stopping the application etc.

So yes, you need to do this "By Hand" [manually] and I deem that a good approach because it means you are explicitly opting in to a "Breaking change" / scheduled downtime / stopping the application style migration.

**Comment by Incanus3** [comment]
hi @rbygrave. yeah, this makes perfect sense and I have no problem with that. I just thought it would be nice to do the manual work in the xml, which is backend agnostic, and then generate the sql files from that.
